### PR TITLE
Add Kotlin Native v1.3.31

### PR DESCRIPTION
**Pull request by grlee**
There's a formula that was submitted and rejected to homebrew-core.  I believe a cask makes more sense, given the nature of the project (e.g. native binaries are included in the release)

After making all changes to the cask:

- [X] `brew cask audit --download {{cask_file}}` is error-free.
- [X] `brew cask style --fix {{cask_file}}` reports no offenses.
- [X] The commit message includes the cask’s name and version.
- [X] The submission is for [a stable version](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask) or [documented exception](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#but-there-is-no-stable-version).

Additionally, if **adding a new cask**:

- [X] Named the cask according to the [token reference].
- [X] `brew cask install {{cask_file}}` worked successfully.
- [X] `brew cask uninstall {{cask_file}}` worked successfully.
- [X] Checked there are no [open pull requests] for the same cask.
- [X] Checked the cask was not [already refused].
- [X] Checked the cask is submitted to [the correct repo].

[token reference]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md
[open pull requests]: https://github.com/Homebrew/homebrew-cask/pulls
[already refused]: https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues
[the correct repo]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask
[version-checksum]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/sha256.md#updating-the-sha256

### CODE REVIEW

File: Casks/kotlin-native.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+require 'fileutils'
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: Please don’t add stuff before the cask definition. No other official cask does it, and it is not necessary.

File: Casks/kotlin-native.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      fun main(args: Array<String>) {
+        println("Hello World!")
+      }
+    EOS
+    system 'kotlinc-native', "#{tmpdir}/test.kt", '-o', "#{tmpdir}/test.kexe"
+  end
+
+  uninstall delete: "#{staged_path}/#{token}"
+
+  caveats depends_on_java '6+'
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: `caveats` need to be together in the same block.

File: Casks/kotlin-native.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  caveats <<~EOS
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: But remove this `caveat`; it is not installation-related.

File: Casks/kotlin-native.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    puts 'Testing kotlinc-native...'
+    File.write "#{tmpdir}/test.kt", <<~EOS
+      fun main(args: Array<String>) {
+        println("Hello World!")
+      }
+    EOS
+    system 'kotlinc-native', "#{tmpdir}/test.kt", '-o', "#{tmpdir}/test.kexe"
+  end
+
+  uninstall delete: "#{staged_path}/#{token}"
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: Unnecessary. Please remove it.

File: Casks/kotlin-native.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  binary "kotlin-native-macos-#{version}/bin/cinterop"
+  binary "kotlin-native-macos-#{version}/bin/jsinterop"
+  binary "kotlin-native-macos-#{version}/bin/klib"
+  binary "kotlin-native-macos-#{version}/bin/konan-lldb"
+  binary "kotlin-native-macos-#{version}/bin/konanc"
+  binary "kotlin-native-macos-#{version}/bin/kotlinc"
+  binary "kotlin-native-macos-#{version}/bin/kotlinc-native"
+  binary "kotlin-native-macos-#{version}/bin/run_konan"
+
+  postflight do
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: Search for “shim script” in the repo. We have a way to do this.

**Review by vitorgalvao — CHANGES_REQUESTED** [review]

**Comment by vitorgalvao** [comment]
> There's a formula that was submitted and rejected to homebrew-core.

Please link me to it.

How is this different from the `kotlin` formula? How does this add value?

**Comment by grlee** [comment]
Thank you for looking at this @vitorgalvao.

Here's the pull request for the formula:
https://github.com/Homebrew/homebrew-core/pull/37546

As far as value added - the kotlin formula only includes the kotlin compiler that outputs jvm and llvm compatible binaries.  kotlin-native is a side-project to add compiler support for native binaries.  It's primarily use case is building android/ios applications or java/net libraries with the same code base.
(Linux, Windows, and Mac OS binaries are also supported by the kotlin–native project.)

File: Casks/kotlin-native.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
+require 'fileutils'
+require 'open3'
+require 'tmpdir'
+tmpdir = Dir.mktmpdir
+
+cask 'kotlin-native' do
+  version '1.3.31'
+  sha256 '964e54649f93d3f269f3753cb7b97739d03c26fe2237a144dcecb92779c36d77'
+
+  # github.com/JetBrains/kotlin/releases/download/v was verified as official when first introduced to the cask
```
Comment (vitorgalvao) [inline Casks/kotlin-native.rb]: Just `github.com/JetBrains/kotlin`

**Comment by vitorgalvao** [comment]
Sounds acceptable. Please proceed with the requested fixes so we can merge.

File: Casks/kotlin-native.rb (hunk 5 above)
Comment (grlee) [reply Casks/kotlin-native.rb]: I'm going to drop this postflight clause.  It's testing that the install succeeds, but I'm not seeing any other casks testing their installations like this.

**Comment by grlee** [comment]
Updated pull request with review suggestions.

**Comment by vitorgalvao** [comment]
Looks great. Thank you.

**Pull request merged**